feat(protocol): Accept spans with empty op

rhcarvalho opened the pull request:
Set a default value in case the op is missing. This is to avoid poor UX
when a span is created via an SDK without op and then the whole
transaction gets dropped without any feedback to the user.

rhcarvalho commented on relay-general/src/store/transactions.rs:
```diff
@@ -395,16 +392,26 @@ mod tests {
+        //         .value()
+        //         .unwrap(),
+        //     "op"
+        // );
```
I need help with this madness :)

@jan-auer or @untitaker, how can I check that the `op` was defaulted?

untitaker replied:
Do this instead: https://github.com/getsentry/relay/blob/ab4cfdc70f53b0fd17f554af9a74378859d68630/relay-general/src/store/normalize/user_agent.rs#L180

Much simpler. You can write `@""` as string initially and then use [insta](https://crates.io/crates/insta) the cli-tool to approve the snapshot.

rhcarvalho replied:
Thanks!

rhcarvalho commented on relay-general/src/store/transactions.rs:
```diff
@@ -153,9 +152,7 @@ impl Processor for TransactionsProcessor {
-        if span.op.value().is_none() {
-            return Err(ProcessingAction::InvalidTransaction("span is missing op"));
-        }
+        span.op.get_or_insert_with(|| "op".to_owned());
```
This is where we set the default value for `op`.
Using `"op".to_owned()` to match the code style in the code base, as suggested by Jan.

rhcarvalho commented on relay-general/src/store/transactions.rs:
```diff
@@ -153,9 +152,7 @@ impl Processor for TransactionsProcessor {
 
-        if span.op.value().is_none() {
-            return Err(ProcessingAction::InvalidTransaction("span is missing op"));
-        }
```
All this got changed because we need a mutable reference to set `op`.

While refactoring over video with Jan, we were able to:
- Extract a common error into a variable;
- Reduce indentation levels by using multiple `let ... match` for traversing the data step by step.

rhcarvalho commented on relay-general/src/store/transactions.rs:
```diff
@@ -373,10 +370,10 @@ mod tests {
         let mut event = Annotated::new(Event {
             ty: Annotated::new(EventType::Transaction),
-            timestamp: Annotated::new(Utc.ymd(2000, 1, 1).and_hms(0, 0, 0)),
+            timestamp: Annotated::new(Utc.ymd(2000, 1, 1).and_hms(0, 0, 10)),
```
Changed this such that the transaction now has non-zero duration.
Zero duration is not the common case, and we don't want to mix up intentions in the test.
This test is about missing `op` in `event.contexts.trace` => `event.contexts.trace = "op"` (default value).

rhcarvalho commented on relay-general/src/store/transactions.rs:
```diff
@@ -395,16 +392,27 @@ mod tests {
+            }
+          }
+        }
+        "###);
```
Thanks Markus for showing me this macro.

`timestamp` and `start_timestamp` are set to match the values in the test, 2000-01-01 00:00:00, though I was still not able to determine why, when the tests run, those two fields become the current timestamp instead -- and the tests fail.

rhcarvalho commented on relay-general/src/store/transactions.rs:
```diff
@@ -674,10 +682,10 @@ mod tests {
         let mut event = Annotated::new(Event {
             ty: Annotated::new(EventType::Transaction),
-            timestamp: Annotated::new(Utc.ymd(2000, 1, 1).and_hms(0, 0, 0)),
+            timestamp: Annotated::new(Utc.ymd(2000, 1, 1).and_hms(0, 0, 10)),
```
Same consideration as above, making the test data more realistic: non-zero duration.

rhcarvalho commented on relay-general/src/store/transactions.rs:
```diff
@@ -695,7 +703,7 @@ mod tests {
             })),
             spans: Annotated::new(vec![Annotated::new(Span {
-                timestamp: Annotated::new(Utc.ymd(2000, 1, 1).and_hms(0, 0, 0)),
+                timestamp: Annotated::new(Utc.ymd(2000, 1, 1).and_hms(0, 0, 10)),
```
Same consideration as above, making the test data more realistic: non-zero duration.

rhcarvalho commented on relay-general/src/store/transactions.rs:
```diff
@@ -705,14 +713,36 @@ mod tests {
+            }
+          ]
+        }
+        "###);
```
Same as above, when the tests run, all timestamps in event reflect the current timestamp, not 2000-01-01 00:00:00 as set in the tests 😕

jan-auer commented on relay-general/src/store/transactions.rs:
```diff
@@ -705,14 +713,36 @@ mod tests {
-        );
+        process_value(
+            &mut event,
+            &mut TransactionsProcessor::new(None),
```
You need to set `processor.now` here to avoid clock drift correction (see other tests):

```rs
let mut processor = TransactionsProcessor::new(None);
processor.now = end;
process_value(&mut event, &mut processor, ProcessingState::root()).unwrap();
```

@untitaker clock drift correction looks extremely fragile. It should use a timestamp of when the event was received, not when the processor runs (`now`). Since we can have arbitrary delays inside Relay due to project config fetching, this can massively skew the clock -- more than what's expected from usual clock skew. This means, the `TransactionsProcessor` should always receive `sent_at` and also `received_at` instead of `now`.

Also, doing clock drift correction with the `end` timestamp probably does more harm than it helps. We cannot assume that the transaction is sent immediately after it's complete (especially if we want to start batching up in the SDK).

untitaker replied:
Your comments wrt capturing `now` outside of the processor and inside of the endpoint make sense.

>We cannot assume that the transaction is sent immediately after it's complete (especially if we want to start batching up in the SDK).

But that's precisely why the `sent_at` parameter exists. It is eventually supposed to be passed by the SDK (we didn't impl that part yet), and is supposed to tell Relay what the SDK thinks is the current time. The difference between `sent_at` and `now/received_at` is our skew.

rhcarvalho replied:
@jan-auer thanks for pointing out the problem. Dubious default behavior. And surprising API, setting `processor.now` to disable clock drift correction.

The documentation of `TransactionsProcessor.now` reads:

```rust
// This is an attribute so we can mock it in testing.
```

Nothing about clock drift. `TransactionsProcessor` documentation also does no mention it's surprising behavior of modifying input timestamps.

I've cut https://github.com/getsentry/relay/issues/535 to track improving/documenting that.


---

As for the discussion around the clock drift implementation, I've cut https://github.com/getsentry/relay/issues/534 so we can move the discussion there.

jan-auer reviewed (changes requested):
Thanks, Rodolfo, especially for the detailed comments. See the below comment for why tests are failing still.

rhcarvalho committed:
feat(relay-general): Accept spans with empty op

Set a default value in case the op is missing. This is to avoid poor UX
when a span is created via an SDK without op and then the whole
transaction gets dropped without any feedback to the user.

jan-auer reviewed (approved):
Thanks!

dashed reviewed (approved):
Looks good to me 👍

rhcarvalho commented on relay-general/src/store/transactions.rs:
```diff
@@ -153,9 +152,7 @@ impl Processor for TransactionsProcessor {
-        if span.op.value().is_none() {
-            return Err(ProcessingAction::InvalidTransaction("span is missing op"));
-        }
+        span.op.get_or_insert_with(|| "default".to_owned());
```
Discussed with the team on what the default value should be and we settled on `"default"`.

We avoided "undefined", "null" and similar terms because they could be confused with an UI error.
We considered how it will appear on searches, `transaction.op:default` seemed better than `transaction.op:op`.

@dashed @untitaker @HazAT
